google/cloud/bigquery_storage_v1/writer.py

```diff
@@ -0,0 +1,412 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from __future__ import division
+
+import itertools
+import logging
+import queue
+import time
+import threading
+from typing import Callable, Optional, Sequence, Tuple
+
+from google.api_core import bidi
+from google.api_core.future import polling as polling_future
+from google.api_core import exceptions
+import google.api_core.retry
+import grpc
+
+from google.cloud.bigquery_storage_v1 import exceptions as bqstorage_exceptions
+from google.cloud.bigquery_storage_v1 import types as gapic_types
+from google.cloud.bigquery_storage_v1.services import big_query_write
+
+_LOGGER = logging.getLogger(__name__)
+_RPC_ERROR_THREAD_NAME = "Thread-OnRpcTerminated"
+
+# _open() takes between 0.25 and 0.4 seconds to be ready. Wait each loop before
+# checking again. This interval was chosen to result in about 3 loops.
+_WRITE_OPEN_INTERVAL = 0.08
+
+# Use a default timeout that is quite long to avoid potential infinite loops,
+# but still work for all expected requests
+_DEFAULT_TIMEOUT = 600
+
+
+def _wrap_as_exception(maybe_exception) -> Exception:
+    """Wrap an object as a Python exception, if needed.
+    Args:
+        maybe_exception (Any): The object to wrap, usually a gRPC exception class.
+    Returns:
+         The argument itself if an instance of ``BaseException``, otherwise
+         the argument represented as an instance of ``Exception`` (sub)class.
+    """
+    if isinstance(maybe_exception, grpc.RpcError):
+        return exceptions.from_grpc_error(maybe_exception)
+    elif isinstance(maybe_exception, BaseException):
+        return maybe_exception
+
+    return Exception(maybe_exception)
+
+
+class AppendRowsStream(object):
+    """A manager object which can append rows to a stream."""
+
+    def __init__(
+        self,
+        client: big_query_write.BigQueryWriteClient,
+        initial_request_template: gapic_types.AppendRowsRequest,
+        metadata: Sequence[Tuple[str, str]] = (),
+    ):
+        """Construct a stream manager.
+
+        Args:
+            client:
+                Client responsible for making requests.
+            initial_request_template:
+                Data to include in the first request sent to the stream. This
+                must contain
+                :attr:`google.cloud.bigquery_storage_v1beta2.types.AppendRowsRequest.write_stream`
```

Comment:
nit: update docs to the v1 references as well, there's v1beta2 references scattered about.

Reply:
Good catch. Done!